tests/unit/stack_test.c

```diff
@@ -0,0 +1,119 @@
+#include "test.h"
+
+#include <alloc.h>
+#include <stack.h>
+
+static void test_push_pop(void)
+{
+    Stack *stack = StackNew(0, free);
+
+    StackPush(stack, xstrdup("1"));
+    StackPush(stack, xstrdup("2"));
+    StackPush(stack, xstrdup("3"));
+
+    char *str1 = StackPop(stack);
+    char *str2 = StackPop(stack);
+    char *str3 = StackPop(stack);
+    assert_int_equal(strcmp(str1, "3"), 0);
+    assert_int_equal(strcmp(str2, "2"), 0);
+    assert_int_equal(strcmp(str3, "1"), 0);
+
+    free(str1);
+    free(str2);
+    free(str3);
+
+    StackDestroy(stack);
+}
+
+static void test_pop_empty_and_push_null(void)
+{
+    Stack *stack = StackNew(1, NULL);
+
+    assert(StackIsEmpty(stack));
+
+    void *i_am_null = StackPop(stack);
+    assert(i_am_null == NULL);
+    StackPush(stack, i_am_null);
+    assert(StackPop(stack) == NULL);
+
+    StackDestroy(stack);
+}
+
+static void test_copy(void)
+{
+    Stack *stack = StackNew(0, free);
+
+    StackPush(stack, xstrdup("1"));
+    StackPush(stack, xstrdup("2"));
+    StackPush(stack, xstrdup("3"));
+
+    Stack *new_stack = StackCopy(stack);
+
+    assert(new_stack != NULL);
+
+    char *old_str1 = StackPop(stack); char *new_str1 = StackPop(new_stack);
+    char *old_str2 = StackPop(stack); char *new_str2 = StackPop(new_stack);
+    char *old_str3 = StackPop(stack); char *new_str3 = StackPop(new_stack);
+
+    assert_int_equal(strcmp(old_str1, new_str1), 0);
+    assert_int_equal(strcmp(old_str2, new_str2), 0);
+    assert_int_equal(strcmp(old_str3, new_str3), 0);
+
+    free(old_str1);
+    free(old_str2);
+    free(old_str3);
+
+    StackSoftDestroy(stack);
+    StackDestroy(new_stack);
```

Comment:
Maybe check the state of `new_stack` between these two lines?

Reply:
That's a really comprehensive check of the stack. :smile: I meant something like checking it's size, capacity, maybe trying to push/pop. And even like this, it should at least be `assert(new_stack != NULL)`.